### ISSUE TITLE
Reachability analysis for `in` is inconsistent with type narrowing

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ty#4040]
### Summary

The first of these does not cause us to emit an `invalid-return-type` error, because we correctly infer that the `==` operation will always evaluate to `True`. The second one causes us to emit a false-positive `invalid-return-type` error, even though `tuple.__contains__` uses equality under the hood:

```py
from typing import Literal

def f(x: Literal["foo"]) -> int:
    if x == "foo":
        return 42

def g(x: Literal["foo"]) -> int:  # `invalid-return-type`
    if x in ("foo",):
        return 42
```

Obviously types can freely override `__contains__`, so inferring a precise `Literal[True]` type for an `in` expression  isn't necessarily safe unless you know the exact class of the r.h.s. operand in the `in` check. But in cases where we feel we're able to confidently narrow the type using an `in` test, I feel like we should try to make our reachability analysis consistent with that by inferring a precise `Literal[True, False]` type for the result of the `in` operation. This is one such case: if you add a `reveal_type` call, the `invalid-return-type` error goes away and ty reveals that the inferred type after the `in` check is `Never`:

```py
from typing import Literal

def g(x: Literal["foo"]) -> int:
    if x in ("foo",):
        return 42
    reveal_type(x)  # Never
```
